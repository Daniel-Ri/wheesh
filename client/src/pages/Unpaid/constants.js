export const GET_ORDER = 'Unpaid/GET_ORDER';
export const SET_ORDER = 'Unpaid/SET_ORDER';
export const CANCEL_ORDER = 'Unpaid/CANCEL_ORDER';
export const PAY_ORDER = 'Unpaid/PAY_ORDER';
