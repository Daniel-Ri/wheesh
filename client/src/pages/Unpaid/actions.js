import { CANCEL_ORDER, GET_ORDER, PAY_ORDER, SET_ORDER } from './constants';

export const getOrder = (orderId, handleError) => ({
  type: GET_ORDER,
  orderId,
  handleError,
});

export const setOrder = (order) => ({
  type: SET_ORDER,
  order,
});

export const cancelOrder = (orderId, handleSuccess, handleError) => ({
  type: CANCEL_ORDER,
  orderId,
  handleSuccess,
  handleError,
});

export const payOrder = (orderId, handleSuccess, handleError) => ({
  type: PAY_ORDER,
  orderId,
  handleSuccess,
  handleError,
});
