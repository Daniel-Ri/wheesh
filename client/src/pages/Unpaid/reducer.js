import { produce } from 'immer';
import { SET_ORDER } from './constants';

export const initialState = {
  order: null,
};

export const storedKey = [];

const unpaidReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case SET_ORDER:
        draft.order = action.order;
        break;
    }
  });

export default unpaidReducer;
