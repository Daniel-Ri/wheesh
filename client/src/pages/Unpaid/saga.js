import { takeLatest, call, put } from 'redux-saga/effects';
import { cancelOrder, getOrder, payOrder } from '@domain/api';
import { CANCEL_ORDER, GET_ORDER, PAY_ORDER } from './constants';
import { setOrder } from './actions';

function* doGetOrder({ orderId, handleError }) {
  try {
    const response = yield call(getOrder, orderId);
    yield put(setOrder(response.data));
  } catch (error) {
    handleError(error.response.data.message);
  }
}

function* doCancelOrder({ orderId, handleSuccess, handleError }) {
  try {
    yield call(cancelOrder, orderId);
    yield call(handleSuccess);
  } catch (error) {
    handleError(error.response.data.message);
  }
}

function* doPayOrder({ orderId, handleSuccess, handleError }) {
  try {
    yield call(payOrder, orderId);
    yield call(handleSuccess);
  } catch (error) {
    handleError(error.response.data.message);
  }
}

export default function* unpaidSaga() {
  yield takeLatest(GET_ORDER, doGetOrder);
  yield takeLatest(CANCEL_ORDER, doCancelOrder);
  yield takeLatest(PAY_ORDER, doPayOrder);
}
