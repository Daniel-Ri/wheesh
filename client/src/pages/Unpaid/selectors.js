import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectUnpaidState = (state) => state.unpaid || initialState;

export const selectOrder = createSelector(selectUnpaidState, (state) => state.order);
