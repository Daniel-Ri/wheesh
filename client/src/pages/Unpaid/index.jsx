import { connect, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { useNavigate, useParams } from 'react-router-dom';

import arrowImage from '@static/images/arrowTrain.png';
import { createStructuredSelector } from 'reselect';
import { useEffect, useState } from 'react';
import toast from 'react-hot-toast';
import BackBtn from '@components/BackBtn';
import { formatDateWithDay, formatHour, formatRupiah } from '@utils/handleValue';
import { Button } from '@mui/material';
import classes from './style.module.scss';
import { cancelOrder, getOrder, payOrder, setOrder } from './actions';
import { selectOrder } from './selectors';
import OrderedSeatCard from './components/OrderedSeatCard';

const Unpaid = ({ order }) => {
  const { orderId } = useParams();
  const [countdown, setCountdown] = useState({ minutes: 0, seconds: 0 });

  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleErrorGetOrder = (errorMsg) => {
    toast.error(errorMsg);
    navigate('/my-tickets');
  };

  const handleErrorCancel = (errorMsg) => {
    toast.error(errorMsg);
  };

  const handleSuccessCancel = () => {
    toast.success('Success cancel the order');
    dispatch(setOrder(null));
    navigate('/my-tickets');
  };

  const handleErrorPay = (errorMsg) => {
    toast.error(errorMsg);
  };

  const handleSuccessPay = () => {
    toast.success('Success pay the order');
    dispatch(setOrder(null));
    navigate(`/order/${orderId}`);
  };

  const handleCancel = () => {
    dispatch(cancelOrder(orderId, handleSuccessCancel, handleErrorCancel));
  };

  const handlePay = () => {
    dispatch(payOrder(orderId, handleSuccessPay, handleErrorPay));
  };

  useEffect(() => {
    dispatch(getOrder(orderId, handleErrorGetOrder));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [orderId]);

  useEffect(() => {
    if (!order) return;

    if (order.Payment.isPaid) {
      toast.error('You have paid the order');
      navigate('/my-tickets');
    }

    const countdowner = setInterval(() => {
      const now = new Date().getTime();
      const duePayment = new Date(order.Payment.duePayment).getTime();

      if (now > duePayment) {
        toast.error('Passed payment due time');
        navigate('/my-tickets');
      }

      const distance = duePayment - now;
      const minutes = Math.floor(distance / (60 * 1000));
      const seconds = Math.floor((distance % (60 * 1000)) / 1000);
      setCountdown({ minutes, seconds });
    }, 1000);

    return () => {
      clearInterval(countdowner);
      dispatch(setOrder(null));
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [order]);

  return (
    <main className={classes.main}>
      <div className={classes.container}>
        <header>
          <BackBtn handleClickBack={() => navigate('/my-tickets')} />
          <h1>Unpaid</h1>
        </header>

        <div className={classes.countdown}>
          <div className={classes.message}>To be paid, remaining time:</div>
          <div className={classes.time}>
            {countdown.minutes} m {countdown.seconds} s
          </div>
        </div>

        <section>
          <div className={classes.sectionDesc}>
            <div className={classes.row}>
              <div className={classes.dateTime}>{formatDateWithDay(order?.Schedule.departureTime)}</div>
            </div>
            <div className={classes.row}>
              <div className={classes.timeAndPlace}>
                <div className={classes.time}>
                  {formatHour(order?.Schedule.departureTime)}
                  <sup> WIB</sup>
                </div>
                <div className={classes.place}>{order?.Schedule.departureStation.name}</div>
              </div>
              <div className={classes.train}>
                <div className={classes.name}>{order?.Schedule.Train.name}</div>
                <div className={classes.arrowImage}>
                  <img src={arrowImage} alt="" />
                </div>
              </div>
              <div className={classes.timeAndPlace}>
                <div className={classes.time}>
                  {formatHour(order?.Schedule.arrivalTime)}
                  <sup> WIB</sup>
                </div>
                <div className={classes.place}>{order?.Schedule.arrivalStation.name}</div>
              </div>
            </div>
          </div>
        </section>

        <section>
          <div className={classes.header}>Passenger</div>
          <hr />
          <div className={classes.sectionDesc}>
            {order?.OrderedSeats.map((orderedSeat, idx) => (
              <OrderedSeatCard
                key={orderedSeat.id}
                orderedSeat={orderedSeat}
                addLine={idx !== order.OrderedSeats.length - 1}
              />
            ))}
          </div>
        </section>

        <section>
          <div className={classes.sectionDesc}>
            <div className={classes.totalPrice}>
              <div>Total Price</div>
              <div>{formatRupiah(order?.Payment.amount)}</div>
            </div>
            <div className={classes.buttons}>
              <Button variant="outlined" className={classes.btn} onClick={handleCancel}>
                Cancel Order
              </Button>
              <Button variant="contained" className={classes.btn} onClick={handlePay}>
                Pay
              </Button>
            </div>
          </div>
        </section>
      </div>
    </main>
  );
};

Unpaid.propTypes = {
  order: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  order: selectOrder,
});

export default connect(mapStateToProps)(Unpaid);
