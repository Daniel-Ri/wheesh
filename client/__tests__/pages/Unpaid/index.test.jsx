import { render, fireEvent } from '@utils/testHelper';
import Unpaid from '@pages/Unpaid';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
  useParams: () => ({
    orderId: 1,
  }),
}));

let wrapper;

const mockProps = {
  order: {
    id: 1,
    userId: 1,
    scheduleId: 91,
    isNotified: false,
    createdAt: '2023-12-14T02:58:48.000Z',
    Payment: {
      id: 1,
      orderId: 1,
      amount: 72000000,
      isPaid: false,
      duePayment: '2023-12-14T02:58:49.000Z',
    },
    Schedule: {
      id: 91,
      trainId: 1,
      departureStationId: 1,
      arrivalStationId: 3,
      departureTime: '2023-12-14T23:15:00.000Z',
      arrivalTime: '2023-12-14T23:45:00.000Z',
      departureStation: {
        id: 1,
        name: 'Halim',
      },
      arrivalStation: {
        id: 3,
        name: 'Padalarang',
      },
      Train: {
        id: 1,
        name: 'G1201',
      },
    },
    OrderedSeats: [
      {
        id: 1,
        orderId: 1,
        seatId: 1,
        price: 600000,
        gender: 'Female',
        dateOfBirth: '2001-07-08T06:10:48.000Z',
        idCard: '0264465324275725',
        name: 'Charlie Doe',
        email: null,
        Seat: {
          id: 1,
          carriageId: 1,
          seatNumber: '1A',
          seatClass: 'first',
          Carriage: {
            id: 1,
            trainId: 1,
            carriageNumber: 1,
          },
        },
      },
      {
        id: 2,
        orderId: 1,
        seatId: 2,
        price: 600000,
        gender: 'Male',
        dateOfBirth: '1977-04-13T04:00:29.000Z',
        idCard: '7032177580189973',
        name: 'Nur Williams',
        email: null,
        Seat: {
          id: 2,
          carriageId: 1,
          seatNumber: '1C',
          seatClass: 'first',
          Carriage: {
            id: 1,
            trainId: 1,
            carriageNumber: 1,
          },
        },
      },
    ],
  },
  locale: 'id',
};

beforeEach(() => {
  wrapper = render(<Unpaid {...mockProps} />);
});

describe('Page Unpaid', () => {
  test('Renders Unpaid page', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('Unpaid')).toBeInTheDocument();
  });

  test('Should call navigate when BackBtn clicked', () => {
    const { getByTestId } = wrapper;
    const backButton = getByTestId('BackBtn');
    fireEvent.click(backButton);
    expect(mockNavigate).toHaveBeenCalledWith('/my-tickets');
  });

  test('Should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
