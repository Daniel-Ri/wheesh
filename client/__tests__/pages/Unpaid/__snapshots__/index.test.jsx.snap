// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Unpaid Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
        data-testid="Unpaid"
      >
        <div
          class="container"
        >
          <header>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="BackBtn"
              tabindex="0"
              type="button"
            >
              Kembali
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <h1>
              Belum Dibayar
            </h1>
          </header>
          <div
            class="countdown"
          >
            <div
              class="message"
            >
              Harus dibayar, sisa waktu:
            </div>
            <div
              class="time"
            >
              0
               m 
              0
               s
            </div>
          </div>
          <section>
            <div
              class="sectionDesc"
            >
              <div
                class="row"
              >
                <div
                  class="dateTime"
                >
                  undefined, NaN Invalid Date NaN
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="timeAndPlace"
                >
                  <div
                    class="time"
                  >
                    Invalid Date
                    <sup>
                       WIB
                    </sup>
                  </div>
                  <div
                    class="place"
                  />
                </div>
                <div
                  class="train"
                >
                  <div
                    class="name"
                  />
                  <div
                    class="arrowImage"
                  >
                    <img
                      alt=""
                      src="/arrowTrain.png"
                    />
                  </div>
                </div>
                <div
                  class="timeAndPlace"
                >
                  <div
                    class="time"
                  >
                    Invalid Date
                    <sup>
                       WIB
                    </sup>
                  </div>
                  <div
                    class="place"
                  />
                </div>
              </div>
            </div>
          </section>
          <section>
            <div
              class="header"
            >
              Penumpang
            </div>
            <hr />
            <div
              class="sectionDesc"
            />
          </section>
          <section>
            <div
              class="sectionDesc"
            >
              <div
                class="totalPrice"
              >
                <div>
                  Harga Total
                </div>
                <div>
                  RpNaN
                </div>
              </div>
              <div
                class="buttons"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium btn css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  Batalkan Pesanan
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  Bayar
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </section>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
      data-testid="Unpaid"
    >
      <div
        class="container"
      >
        <header>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="BackBtn"
            tabindex="0"
            type="button"
          >
            Kembali
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <h1>
            Belum Dibayar
          </h1>
        </header>
        <div
          class="countdown"
        >
          <div
            class="message"
          >
            Harus dibayar, sisa waktu:
          </div>
          <div
            class="time"
          >
            0
             m 
            0
             s
          </div>
        </div>
        <section>
          <div
            class="sectionDesc"
          >
            <div
              class="row"
            >
              <div
                class="dateTime"
              >
                undefined, NaN Invalid Date NaN
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="timeAndPlace"
              >
                <div
                  class="time"
                >
                  Invalid Date
                  <sup>
                     WIB
                  </sup>
                </div>
                <div
                  class="place"
                />
              </div>
              <div
                class="train"
              >
                <div
                  class="name"
                />
                <div
                  class="arrowImage"
                >
                  <img
                    alt=""
                    src="/arrowTrain.png"
                  />
                </div>
              </div>
              <div
                class="timeAndPlace"
              >
                <div
                  class="time"
                >
                  Invalid Date
                  <sup>
                     WIB
                  </sup>
                </div>
                <div
                  class="place"
                />
              </div>
            </div>
          </div>
        </section>
        <section>
          <div
            class="header"
          >
            Penumpang
          </div>
          <hr />
          <div
            class="sectionDesc"
          />
        </section>
        <section>
          <div
            class="sectionDesc"
          >
            <div
              class="totalPrice"
            >
              <div>
                Harga Total
              </div>
              <div>
                RpNaN
              </div>
            </div>
            <div
              class="buttons"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium btn css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Batalkan Pesanan
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Bayar
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </section>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
